connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginsTest.java

```diff
@@ -185,6 +181,102 @@ public void shouldInstantiateAndConfigureDefaultHeaderConverter() {
         assertTrue(headerConverter instanceof SimpleHeaderConverter);
     }
 
+    @Test(expected = ConnectException.class)
+    public void shouldThrowIfPluginThrows() {
+        TestPlugins.assertInitialized();
+
+        plugins.newPlugin(
+            TestPlugins.ALWAYS_THROW_EXCEPTION,
+            new AbstractConfig(new ConfigDef(), Collections.emptyMap()),
+            Converter.class
+        );
+    }
+
+    @Test
+    public void shouldShareStaticValuesBetweenSamePlugin() {
+        // Plugins are not isolated from other instances of their own class.
+        TestPlugins.assertInitialized();
+        Converter firstPlugin = plugins.newPlugin(
+            TestPlugins.SAMPLING,
+            new AbstractConfig(new ConfigDef(), Collections.emptyMap()),
+            Converter.class
+        );
+
+        assertTrue(firstPlugin instanceof SamplingTestPlugin);
+
+        assertEquals(1, ((SamplingTestPlugin) firstPlugin).dynamicInitializations());
+
+        Converter secondPlugin = plugins.newPlugin(
+            TestPlugins.SAMPLING,
+            new AbstractConfig(new ConfigDef(), Collections.emptyMap()),
+            Converter.class
+        );
+
+        assertTrue(secondPlugin instanceof SamplingTestPlugin);
+
+        assertEquals(2, ((SamplingTestPlugin) secondPlugin).dynamicInitializations());
+
+        // This value changes because secondPlugin's instantiation incremented it.
+        assertEquals(2, ((SamplingTestPlugin) firstPlugin).dynamicInitializations());
+    }
+
+    @Test
+    public void shouldPerformServiceLoadingWithPluginClassloader() {
+        TestPlugins.assertInitialized();
+        Converter plugin = plugins.newPlugin(
+            TestPlugins.SERVICE_LOADER,
+            new AbstractConfig(new ConfigDef(), Collections.emptyMap()),
+            Converter.class
+        );
+
+        assertTrue(plugin instanceof SamplingTestPlugin);
+        SamplingTestPlugin samplingPlugin = (SamplingTestPlugin) plugin;
+        for (Entry<String, SamplingTestPlugin> e : samplingPlugin.otherSamples().entrySet()) {
+            String message = e.getKey() + "was not initialized properly";
+            SamplingTestPlugin sample = e.getValue();
+            assertTrue(message, sample.staticClassloader() instanceof PluginClassLoader);
+            assertTrue(message, sample.classloader() instanceof PluginClassLoader);
+            assertEquals(message, sample.staticClassloader(), sample.classloader());
+        }
+        Map<String, SamplingTestPlugin> samples = samplingPlugin.otherSamples();
+        // Initialized once per (static, dynamic, static subclass, dynamic subclass)
```

Comment:
Is it necessary for this test to ensure the exact number of initializations for the classes here? What would we gain from that (what bugs could we proactively catch) that we wouldn't get from just asserting that the number of (successful) initializations is at least one?

Reply:
This is ensuring that all of the initializations happened, indicating that the service loader operated correctly. If you removed one of the service loader calls, or swallowed an exception, this test wouldn't be checking all of the cases. We know that won't happen, but then we're relying on the exception handling to cause the test to fail.